File: R/merge.R
#' @importFrom rlang abort
#' @importFrom methods is
#' @include 00_classes.R
NULL

# ==============================================================================
# DeltaMergeBuilder S7 Class
# ==============================================================================

#' DeltaMergeBuilder S7 Class
#'
#' An S7 class representing a Delta Lake merge operation builder.
#' All configuration is stored in R and only passed to Rust on execute().
#'
#' This class is typically created by calling \code{\link{delta_merge}} rather
#' than constructing it directly.
#'
#' @param table_path Character. Path to the Delta table.
#' @param storage_options Named list or NULL. Storage backend options.
#' @param source_data The source data for the merge operation.
#' @param predicate Character. SQL-like predicate for matching rows.
#' @param source_alias Character. Alias for source table in expressions.
#' @param target_alias Character. Alias for target table in expressions.
#' @param matched_update_clauses List. WHEN MATCHED UPDATE clauses.
#' @param matched_delete_clauses List. WHEN MATCHED DELETE clauses.
#' @param not_matched_insert_clauses List. WHEN NOT MATCHED INSERT clauses.
#' @param not_matched_by_source_update_clauses List. WHEN NOT MATCHED BY SOURCE UPDATE clauses.
#' @param not_matched_by_source_delete_clauses List. WHEN NOT MATCHED BY SOURCE DELETE clauses.
#'
#' @seealso \code{\link{delta_merge}} for creating merge operations.
#'
#' @export
DeltaMergeBuilder <- new_class(
  "DeltaMergeBuilder",
  properties = list(
    # Target table info
    table_path = new_property(class_character),
    storage_options = new_property(class_list | NULL, default = NULL),

    # Source data (kept as-is, converted to Arrow stream on execute)
    source_data = new_property(class_any),

    # Merge configuration
    predicate = new_property(class_character),
    source_alias = new_property(class_character, default = "source"),
    target_alias = new_property(class_character, default = "target"),

    # Clause lists - each element is a list with 'predicate' and/or 'updates'
    matched_update_clauses = new_property(class_list, default = list()),
    matched_delete_clauses = new_property(class_list, default = list()),
    not_matched_insert_clauses = new_property(class_list, default = list()),
    not_matched_by_source_update_clauses = new_property(
      class_list,
      default = list()
    ),
    not_matched_by_source_delete_clauses = new_property(
      class_list,
      default = list()
    )
  )
)

# Print method for DeltaMergeBuilder
method(print, DeltaMergeBuilder) <- function(x, ...) {
  cat("DeltaMergeBuilder\n")
  cat("  Table:", x@table_path, "\n")
  cat("  Predicate:", x@predicate, "\n")
  cat("  Source alias:", x@source_alias, "\n")
  cat("  Target alias:", x@target_alias, "\n")
  cat("  Clauses:\n")
  cat("    - matched_update:", length(x@matched_update_clauses), "\n")
  cat("    - matched_delete:", length(x@matched_delete_clauses), "\n")
  cat("    - not_matched_insert:", length(x@not_matched_insert_clauses), "\n")
  cat(
    "    - not_matched_by_source_update:",
    length(x@not_matched_by_source_update_clauses),
    "\n"
  )
  cat(
    "    - not_matched_by_source_delete:",
    length(x@not_matched_by_source_delete_clauses),
    "\n"
  )
  invisible(x)
}

# ==============================================================================
# Main Entry Point
# ==============================================================================

#' Start a Delta Lake MERGE operation
#'
#' Creates a DeltaMergeBuilder to configure and execute a MERGE operation
#' that can update, insert, or delete records based on matching conditions.
#'
#' @param table A DeltaTable object or character path to Delta table.
#' @param source Source data (data.frame, Arrow Table, or any nanoarrow-compatible object).
#' @param predicate Character. SQL-like predicate for matching (e.g., "target.id = source.id").
#' @param source_alias Character. Alias for source table in predicates (default: "source").
#' @param target_alias Character. Alias for target table in predicates (default: "target").
#' @param storage_options Named list. Storage backend options (optional).
#'
#' @return A DeltaMergeBuilder object that can be further configured with
#'   `when_matched_*` and `when_not_matched_*` methods.
#'
#' @examples
#' \dontrun{
#' # Create target table
#' target <- data.frame(id = 1:3, value = c(10, 20, 30))
#' write_deltalake(target, "path/to/table")
#'
#' # Prepare source data for merge
#' source <- data.frame(id = c(2, 4), value = c(25, 40))
#'
#' # Perform upsert (update existing, insert new)
#' result <- delta_merge("path/to/table", source, "target.id = source.id") |>
#'   when_matched_update(c(value = "source.value")) |>
#'   when_not_matched_insert(c(id = "source.id", value = "source.value")) |>
#'   merge_execute()
#' }
#'
#' @seealso
#' \code{\link{when_matched_update}}, \code{\link{when_matched_delete}},
#' \code{\link{when_not_matched_insert}}, \code{\link{when_not_matched_by_source_delete}},
#' \code{\link{merge_execute}}
#'
#' @export
delta_merge <- function(
  table,
  source,
  predicate,
  source_alias = "source",
  target_alias = "target",
  storage_options = NULL
) {
  # Validate inputs
  if (
    !is.character(predicate) || length(predicate) != 1 || nchar(predicate) == 0
  ) {
    stop("'predicate' must be a non-empty character string")
  }

  if (!is.character(source_alias) || length(source_alias) != 1) {
    stop("'source_alias' must be a single character string")
  }

  if (!is.character(target_alias) || length(target_alias) != 1) {
    stop("'target_alias' must be a single character string")
  }

  # Extract path from DeltaTable or use as-is
  # S7 classes use S7_class() for type checking
  table_path <- if (S7::S7_inherits(table, DeltaTable)) {
    table@path
  } else if (is.character(table) && length(table) == 1) {
    table
  } else {
    stop("'table' must be a DeltaTable object or a single character path")
  }

  DeltaMergeBuilder(
    table_path = table_path,
    storage_options = storage_options,
    source_data = source,
    predicate = predicate,
    source_alias = source_alias,
    target_alias = target_alias
  )
}

# ==============================================================================
# WHEN MATCHED Clauses
# ==============================================================================

#' Add WHEN MATCHED THEN UPDATE clause
#'
#' Specifies how to update target rows when they match source rows.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{updates}{Named character vector mapping target columns to source expressions.
#'     Names are target column names, values are SQL expressions (e.g., "source.column").}
#'   \item{predicate}{Optional character. Additional predicate to filter which matched
#'     rows should be updated.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @examples
#' \dontrun{
#' # Update all matched rows
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_matched_update(c(value = "source.value", name = "source.name"))
#'
#' # Update only matched rows where a condition is met
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_matched_update(
#'     c(value = "source.value"),
#'     predicate = "source.value > target.value"
#'   )
#' }
#'
#' @export
when_matched_update <- new_generic(
  "when_matched_update",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_matched_update, DeltaMergeBuilder) <- function(
  builder,
  updates,
  predicate = NULL
) {
  if (is.null(updates) || length(updates) == 0) {
    stop("'updates' must be a non-empty named vector")
  }

  if (is.null(names(updates)) || any(names(updates) == "")) {
    stop("'updates' must be a named vector with column names")
  }

  clause <- list(updates = as.list(updates))
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@matched_update_clauses <- c(
    builder@matched_update_clauses,
    list(clause)
  )
  builder
}

#' Add WHEN MATCHED THEN UPDATE ALL clause
#'
#' Updates all columns in matched target rows with corresponding source values.
#' Requires source and target to have matching column names.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{predicate}{Optional character. Additional predicate to filter which matched
#'     rows should be updated.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @note This function is not yet fully supported and will error on execution.
#'   Please use \code{\link{when_matched_update}} with explicit column mappings instead.
#'
#' @examples
#' \dontrun{
#' # Update all columns for matched rows (NOT YET SUPPORTED)
#' # Use when_matched_update with explicit columns instead:
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_matched_update(c(col1 = "source.col1", col2 = "source.col2"))
#' }
#'
#' @export
when_matched_update_all <- new_generic(
  "when_matched_update_all",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_matched_update_all, DeltaMergeBuilder) <- function(
  builder,
  predicate = NULL
) {
  warning(
    "when_matched_update_all() is not yet fully supported. ",
    "Please use when_matched_update() with explicit column mappings instead.",
    call. = FALSE
  )

  clause <- list(update_all = TRUE)
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@matched_update_clauses <- c(
    builder@matched_update_clauses,
    list(clause)
  )
  builder
}

#' Add WHEN MATCHED THEN DELETE clause
#'
#' Deletes target rows that match source rows.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{predicate}{Optional character. Additional predicate to filter which matched
#'     rows should be deleted.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @examples
#' \dontrun{
#' # Delete all matched rows
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_matched_delete()
#'
#' # Delete matched rows where a condition is met
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_matched_delete(predicate = "source.deleted = true")
#' }
#'
#' @export
when_matched_delete <- new_generic(
  "when_matched_delete",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_matched_delete, DeltaMergeBuilder) <- function(
  builder,
  predicate = NULL
) {
  clause <- list()
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@matched_delete_clauses <- c(
    builder@matched_delete_clauses,
    list(clause)
  )
  builder
}

# ==============================================================================
# WHEN NOT MATCHED Clauses (source rows not in target)
# ==============================================================================

#' Add WHEN NOT MATCHED THEN INSERT clause
#'
#' Inserts source rows that don't match any target rows.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{updates}{Named character vector mapping target columns to source expressions.
#'     Names are target column names, values are SQL expressions (e.g., "source.column").}
#'   \item{predicate}{Optional character. Additional predicate to filter which non-matched
#'     source rows should be inserted.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @examples
#' \dontrun{
#' # Insert non-matched source rows
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_insert(c(id = "source.id", value = "source.value"))
#'
#' # Insert with a condition
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_insert(
#'     c(id = "source.id", value = "source.value"),
#'     predicate = "source.value IS NOT NULL"
#'   )
#' }
#'
#' @export
when_not_matched_insert <- new_generic(
  "when_not_matched_insert",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_not_matched_insert, DeltaMergeBuilder) <- function(
  builder,
  updates,
  predicate = NULL
) {
  if (is.null(updates) || length(updates) == 0) {
    stop("'updates' must be a non-empty named vector")
  }

  if (is.null(names(updates)) || any(names(updates) == "")) {
    stop("'updates' must be a named vector with column names")
  }

  clause <- list(updates = as.list(updates))
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@not_matched_insert_clauses <- c(
    builder@not_matched_insert_clauses,
    list(clause)
  )
  builder
}

#' Add WHEN NOT MATCHED THEN INSERT ALL clause
#'
#' Inserts all columns for source rows that don't match any target rows.
#' Requires source and target to have matching column names.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{predicate}{Optional character. Additional predicate to filter which non-matched
#'     source rows should be inserted.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @note This function is not yet fully supported and will error on execution.
#'   Please use \code{\link{when_not_matched_insert}} with explicit column mappings instead.
#'
#' @examples
#' \dontrun{
#' # Insert all columns for non-matched source rows (NOT YET SUPPORTED)
#' # Use when_not_matched_insert with explicit columns instead:
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_insert(c(col1 = "source.col1", col2 = "source.col2"))
#' }
#'
#' @export
when_not_matched_insert_all <- new_generic(
  "when_not_matched_insert_all",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_not_matched_insert_all, DeltaMergeBuilder) <- function(
  builder,
  predicate = NULL
) {
  warning(
    "when_not_matched_insert_all() is not yet fully supported. ",
    "Please use when_not_matched_insert() with explicit column mappings instead.",
    call. = FALSE
  )

  clause <- list(insert_all = TRUE)
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@not_matched_insert_clauses <- c(
    builder@not_matched_insert_clauses,
    list(clause)
  )
  builder
}

# ==============================================================================
# WHEN NOT MATCHED BY SOURCE Clauses (target rows not in source)
# ==============================================================================

#' Add WHEN NOT MATCHED BY SOURCE THEN UPDATE clause
#'
#' Updates target rows that don't have a match in the source data.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{updates}{Named character vector mapping target columns to expressions.
#'     Names are target column names, values are SQL expressions.}
#'   \item{predicate}{Optional character. Additional predicate to filter which non-matched
#'     target rows should be updated.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @examples
#' \dontrun{
#' # Mark unmatched target rows as inactive
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_by_source_update(c(active = "false"))
#' }
#'
#' @export
when_not_matched_by_source_update <- new_generic(
  "when_not_matched_by_source_update",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_not_matched_by_source_update, DeltaMergeBuilder) <- function(
  builder,
  updates,
  predicate = NULL
) {
  if (is.null(updates) || length(updates) == 0) {
    stop("'updates' must be a non-empty named vector")
  }

  if (is.null(names(updates)) || any(names(updates) == "")) {
    stop("'updates' must be a named vector with column names")
  }

  clause <- list(updates = as.list(updates))
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@not_matched_by_source_update_clauses <- c(
    builder@not_matched_by_source_update_clauses,
    list(clause)
  )
  builder
}

#' Add WHEN NOT MATCHED BY SOURCE THEN DELETE clause
#'
#' Deletes target rows that don't have a match in the source data.
#' This is useful for syncing a target table to exactly match the source.
#'
#' @param builder A DeltaMergeBuilder object.
#' @param ... Additional arguments passed to methods.
#'
#' @section Method arguments:
#' The method for DeltaMergeBuilder accepts:
#' \describe{
#'   \item{predicate}{Optional character. Additional predicate to filter which non-matched
#'     target rows should be deleted.}
#' }
#'
#' @return The modified DeltaMergeBuilder (for method chaining).
#'
#' @examples
#' \dontrun{
#' # Delete target rows that are not in source
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_by_source_delete()
#'
#' # Delete with a condition
#' delta_merge(table, source, "target.id = source.id") |>
#'   when_not_matched_by_source_delete(predicate = "target.deletable = true")
#' }
#'
#' @export
when_not_matched_by_source_delete <- new_generic(
  "when_not_matched_by_source_delete",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(when_not_matched_by_source_delete, DeltaMergeBuilder) <- function(
  builder,
  predicate = NULL
) {
  clause <- list()
  if (!is.null(predicate)) {
    if (!is.character(predicate) || length(predicate) != 1) {
      stop("'predicate' must be a single character string")
    }
    clause$predicate <- predicate
  }

  builder@not_matched_by_source_delete_clauses <- c(
    builder@not_matched_by_source_delete_clauses,
    list(clause)
  )
  builder
}

# ==============================================================================
# Execute Method
# ==============================================================================

#' Execute the MERGE operation
#'
#' Executes the configured merge operation against the target Delta table.
#'
#' @param builder A DeltaMergeBuilder object configured with merge clauses.
#' @param ... Additional arguments passed to methods.
#'
#' @return A named list with merge metrics:
#'   \itemize{
#'     \item \code{num_target_rows_inserted}: Number of rows inserted into target.
#'     \item \code{num_target_rows_updated}: Number of rows updated in target.
#'     \item \code{num_target_rows_deleted}: Number of rows deleted from target.
#'     \item \code{num_target_files_added}: Number of files added.
#'     \item \code{num_target_files_removed}: Number of files removed.
#'     \item \code{num_target_rows_copied}: Number of rows copied (unchanged).
#'     \item \code{num_output_rows}: Total number of rows in output.
#'     \item \code{execution_time_ms}: Execution time in milliseconds.
#'   }
#'
#' @examples
#' \dontrun{
#' result <- delta_merge("path/to/table", source, "target.id = source.id") |>
#'   when_matched_update(c(value = "source.value")) |>
#'   when_not_matched_insert(c(id = "source.id", value = "source.value")) |>
#'   merge_execute()
#'
#' print(result)
#' # $num_target_rows_updated
#' # [1] 5
#' # $num_target_rows_inserted
#' # [1] 3
#' # ...
#' }
#'
#' @export
merge_execute <- new_generic(
  "merge_execute",
  "builder",
  function(builder, ...) {
    S7::S7_dispatch()
  }
)

method(merge_execute, DeltaMergeBuilder) <- function(builder) {
  # Validate that at least one clause is defined
  total_clauses <- length(builder@matched_update_clauses) +
    length(builder@matched_delete_clauses) +
    length(builder@not_matched_insert_clauses) +
    length(builder@not_matched_by_source_update_clauses) +
    length(builder@not_matched_by_source_delete_clauses)

  if (total_clauses == 0) {
    stop(
      "At least one merge clause must be specified (e.g., when_matched_update, when_not_matched_insert)"
    )
  }

  # Convert source data to Arrow stream
  stream <- nanoarrow::as_nanoarrow_array_stream(builder@source_data)

  # Call single Rust function with all configuration
  result <- delta_merge_execute(
    table_uri = builder@table_path,
    source_stream = stream,
    predicate = builder@predicate,
    source_alias = builder@source_alias,
    target_alias = builder@target_alias,
    matched_update_clauses = builder@matched_update_clauses,
    matched_delete_clauses = builder@matched_delete_clauses,
    not_matched_insert_clauses = builder@not_matched_insert_clauses,
    not_matched_by_source_update_clauses = builder@not_matched_by_source_update_clauses,
    not_matched_by_source_delete_clauses = builder@not_matched_by_source_delete_clauses,
    storage_options = builder@storage_options
  )

  if (methods::is(result, "error")) {
    rlang::abort(result$value)
  }

  result
}

File: src/rust/src/merge.rs
//! MERGE support for Delta Lake tables
//!
//! This module provides functionality to merge data into Delta Lake tables
//! using delta-rs MergeBuilder. The merge configuration is built entirely
//! in R and passed to a single Rust function for execution.

use std::collections::HashMap;
use std::sync::Arc;

use arrow::ffi_stream::ArrowArrayStreamReader;
use arrow_extendr::from::FromArrowRobj;
use deltalake::arrow::array::RecordBatch;
use deltalake::arrow::array::RecordBatchReader;
use deltalake::datafusion::datasource::MemTable;
use deltalake::datafusion::prelude::*;
use deltalake::operations::merge::MergeBuilder;
use deltalake::DeltaTable;
use extendr_api::prelude::*;

use crate::{block_on, parse_storage_options, path_to_url};

/// Execute a Delta Lake MERGE operation
///
/// This function receives all merge configuration from R and executes
/// the merge in a single call, avoiding complex state management.
///
/// @param table_uri Path to the Delta table
/// @param source_stream Arrow data stream for source data
/// @param predicate Main merge predicate (e.g., "target.id = source.id")
/// @param source_alias Alias for source table in expressions
/// @param target_alias Alias for target table in expressions
/// @param matched_update_clauses List of update clauses for matched rows
/// @param matched_delete_clauses List of delete clauses for matched rows
/// @param not_matched_insert_clauses List of insert clauses for unmatched source rows
/// @param not_matched_by_source_update_clauses List of update clauses for unmatched target rows
/// @param not_matched_by_source_delete_clauses List of delete clauses for unmatched target rows
/// @param storage_options Storage backend options (optional)
#[extendr]
pub fn delta_merge_execute(
    table_uri: &str,
    source_stream: Robj,
    predicate: &str,
    source_alias: &str,
    target_alias: &str,
    matched_update_clauses: List,
    matched_delete_clauses: List,
    not_matched_insert_clauses: List,
    not_matched_by_source_update_clauses: List,
    not_matched_by_source_delete_clauses: List,
    storage_options: Nullable<List>,
) -> Result<List> {
    // Convert R Arrow stream to reader
    let reader = ArrowArrayStreamReader::from_arrow_robj(&source_stream)
        .map_err(|e| Error::from(format!("Failed to read Arrow stream: {:?}", e)))?;

    let schema = reader.schema();
    let boxed_reader: Box<dyn RecordBatchReader + Send + 'static> = Box::new(reader);

    // Parse URL and open table
    let url = path_to_url(table_uri).map_err(Error::from)?;

    let mut table = block_on(async {
        match storage_options {
            Nullable::NotNull(ref opts) => {
                let options = parse_storage_options(opts);
                DeltaTable::try_from_url_with_storage_options(url.clone(), options).await
            }
            Nullable::Null => DeltaTable::try_from_url(url.clone()).await,
        }
    })
    .map_err(|e| Error::from(e.to_string()))?;

    // Load the table
    block_on(async { table.load().await }).map_err(|e| Error::from(e.to_string()))?;

    // Collect all record batches from the source stream into memory
    let mut batches: Vec<RecordBatch> = Vec::new();
    let mut reader_box = boxed_reader;
    while let Some(batch_result) = reader_box.next() {
        let batch =
            batch_result.map_err(|e| Error::from(format!("Failed to read batch: {}", e)))?;
        batches.push(batch);
    }

    // Create a MemTable from the collected batches
    let mem_table = MemTable::try_new(schema.clone(), vec![batches])
        .map_err(|e| Error::from(format!("Failed to create memory table: {}", e)))?;

    // Create SessionContext and register the source table
    let ctx = SessionContext::new();
    block_on(async { ctx.register_table(source_alias, Arc::new(mem_table)) })
        .map_err(|e| Error::from(format!("Failed to register source table: {}", e)))?;

    // Get the source as a DataFrame
    let source_df = block_on(async { ctx.table(source_alias).await })
        .map_err(|e| Error::from(format!("Failed to get source DataFrame: {}", e)))?;

    // Get the table state - the table must be loaded
    let table_state = table
        .state
        .as_ref()
        .ok_or_else(|| Error::from("Table must be loaded before merge"))?;

    // Build merge operation - MergeBuilder::new takes Option<EagerSnapshot>
    let mut merge_builder = MergeBuilder::new(
        table.log_store(),
        Some(table_state.snapshot().clone()),
        predicate.to_string(),
        source_df,
    )
    .with_source_alias(source_alias)
    .with_target_alias(target_alias);

    // Add WHEN MATCHED UPDATE clauses
    for clause in matched_update_clauses.iter() {
        if let Some(clause_list) = clause.1.as_list() {
            merge_builder = add_matched_update_clause(merge_builder, clause_list)?;
        }
    }

    // Add WHEN MATCHED DELETE clauses
    for clause in matched_delete_clauses.iter() {
        if let Some(clause_list) = clause.1.as_list() {
            merge_builder = add_matched_delete_clause(merge_builder, clause_list)?;
        }
    }

    // Add WHEN NOT MATCHED INSERT clauses
    for clause in not_matched_insert_clauses.iter() {
        if let Some(clause_list) = clause.1.as_list() {
            merge_builder = add_not_matched_insert_clause(merge_builder, clause_list)?;
        }
    }

    // Add WHEN NOT MATCHED BY SOURCE UPDATE clauses
    for clause in not_matched_by_source_update_clauses.iter() {
        if let Some(clause_list) = clause.1.as_list() {
            merge_builder = add_not_matched_by_source_update_clause(merge_builder, clause_list)?;
        }
    }

    // Add WHEN NOT MATCHED BY SOURCE DELETE clauses
    for clause in not_matched_by_source_delete_clauses.iter() {
        if let Some(clause_list) = clause.1.as_list() {
            merge_builder = add_not_matched_by_source_delete_clause(merge_builder, clause_list)?;
        }
    }

    // Execute merge
    let (_table, metrics) = block_on(async { merge_builder.await })
        .map_err(|e| Error::from(format!("Merge failed: {}", e)))?;

    // Return metrics as R list
    Ok(list!(
        num_target_rows_inserted = metrics.num_target_rows_inserted as i64,
        num_target_rows_updated = metrics.num_target_rows_updated as i64,
        num_target_rows_deleted = metrics.num_target_rows_deleted as i64,
        num_target_files_added = metrics.num_target_files_added as i64,
        num_target_files_removed = metrics.num_target_files_removed as i64,
        num_target_rows_copied = metrics.num_target_rows_copied as i64,
        num_output_rows = metrics.num_output_rows as i64,
        execution_time_ms = metrics.execution_time_ms as i64
    ))
}

// ============================================================================
// Helper Functions to Add Clauses
// ============================================================================

/// Extract updates map from an R list
fn extract_updates(clause: &List) -> Result<HashMap<String, String>> {
    let mut updates = HashMap::new();

    // Try to get "updates" key from the list by iterating
    for (key, value) in clause.iter() {
        if key == "updates" {
            if let Some(updates_list) = value.as_list() {
                for (col_key, col_value) in updates_list.iter() {
                    if let Some(v) = col_value.as_str() {
                        updates.insert(col_key.to_string(), v.to_string());
                    }
                }
            }
            break;
        }
    }

    Ok(updates)
}

/// Extract optional predicate from an R list
fn extract_predicate(clause: &List) -> Option<String> {
    for (key, value) in clause.iter() {
        if key == "predicate" {
            return value.as_str().map(|s| s.to_string());
        }
    }
    None
}

/// Check if this is an "update_all" clause
fn is_update_all(clause: &List) -> bool {
    for (key, value) in clause.iter() {
        if key == "update_all" {
            return value.as_bool().unwrap_or(false);
        }
    }
    false
}

/// Check if this is an "insert_all" clause
fn is_insert_all(clause: &List) -> bool {
    for (key, value) in clause.iter() {
        if key == "insert_all" {
            return value.as_bool().unwrap_or(false);
        }
    }
    false
}

/// Add a WHEN MATCHED UPDATE clause to the builder
fn add_matched_update_clause(builder: MergeBuilder, clause: List) -> Result<MergeBuilder> {
    let predicate = extract_predicate(&clause);

    if is_update_all(&clause) {
        // UPDATE ALL - we need to get target schema and build updates for all columns
        // Since update_all isn't directly supported, we skip this for now
        // and require explicit column mappings
        return Err(Error::from(
            "when_matched_update_all is not yet supported. Please use when_matched_update with explicit column mappings."
        ));
    }

    // UPDATE with specific columns
    let updates = extract_updates(&clause)?;
    let result = match predicate {
        Some(p) => builder
            .when_matched_update(|mut update| {
                update = update.predicate(p);
                for (col, expr) in updates {
                    update = update.update(col, expr);
                }
                update
            })
            .map_err(|e| Error::from(e.to_string()))?,
        None => builder
            .when_matched_update(|mut update| {
                for (col, expr) in updates {
                    update = update.update(col, expr);
                }
                update
            })
            .map_err(|e| Error::from(e.to_string()))?,
    };
    Ok(result)
}

/// Add a WHEN MATCHED DELETE clause to the builder
fn add_matched_delete_clause(builder: MergeBuilder, clause: List) -> Result<MergeBuilder> {
    let predicate = extract_predicate(&clause);

    let result = match predicate {
        Some(p) => builder
            .when_matched_delete(|delete| delete.predicate(p))
            .map_err(|e| Error::from(e.to_string()))?,
        None => builder
            .when_matched_delete(|delete| delete)
            .map_err(|e| Error::from(e.to_string()))?,
    };
    Ok(result)
}

/// Add a WHEN NOT MATCHED INSERT clause to the builder
fn add_not_matched_insert_clause(builder: MergeBuilder, clause: List) -> Result<MergeBuilder> {
    let predicate = extract_predicate(&clause);

    if is_insert_all(&clause) {
        // INSERT ALL - not directly supported, require explicit mappings
        return Err(Error::from(
            "when_not_matched_insert_all is not yet supported. Please use when_not_matched_insert with explicit column mappings."
        ));
    }

    // INSERT with specific columns
    let updates = extract_updates(&clause)?;
    let result = match predicate {
        Some(p) => builder
            .when_not_matched_insert(|mut insert| {
                insert = insert.predicate(p);
                for (col, expr) in updates {
                    insert = insert.set(col, expr);
                }
                insert
            })
            .map_err(|e| Error::from(e.to_string()))?,
        None => builder
            .when_not_matched_insert(|mut insert| {
                for (col, expr) in updates {
                    insert = insert.set(col, expr);
                }
                insert
            })
            .map_err(|e| Error::from(e.to_string()))?,
    };
    Ok(result)
}

/// Add a WHEN NOT MATCHED BY SOURCE UPDATE clause to the builder
fn add_not_matched_by_source_update_clause(
    builder: MergeBuilder,
    clause: List,
) -> Result<MergeBuilder> {
    let predicate = extract_predicate(&clause);
    let updates = extract_updates(&clause)?;

    let result = match predicate {
        Some(p) => builder
            .when_not_matched_by_source_update(|mut update| {
                update = update.predicate(p);
                for (col, expr) in updates {
                    update = update.update(col, expr);
                }
                update
            })
            .map_err(|e| Error::from(e.to_string()))?,
        None => builder
            .when_not_matched_by_source_update(|mut update| {
                for (col, expr) in updates {
                    update = update.update(col, expr);
                }
                update
            })
            .map_err(|e| Error::from(e.to_string()))?,
    };
    Ok(result)
}

/// Add a WHEN NOT MATCHED BY SOURCE DELETE clause to the builder
fn add_not_matched_by_source_delete_clause(
    builder: MergeBuilder,
    clause: List,
) -> Result<MergeBuilder> {
    let predicate = extract_predicate(&clause);

    let result = match predicate {
        Some(p) => builder
            .when_not_matched_by_source_delete(|delete| delete.predicate(p))
            .map_err(|e| Error::from(e.to_string()))?,
        None => builder
            .when_not_matched_by_source_delete(|delete| delete)
            .map_err(|e| Error::from(e.to_string()))?,
    };
    Ok(result)
}

// Export module
extendr_module! {
    mod merge;
    fn delta_merge_execute;
}
